### PR TITLE
shadps4: init at 0.3.0-unstable-2024-10-13

### LINKED ISSUE #339617: Package request: shadps4
**Project description**

shadPS4 is a cross platform PlayStation 4 emulator.

**Metadata**

* homepage URL: https://shadps4.net/
* source URL: https://github.com/shadps4-emu/shadPS4
* license: gpl-2.0
* platforms: linux, darwin, windows

### PR DESCRIPTION
**Pull request by ryand56**
## Description of changes

Adds [shadPS4](https://github.com/shadps4-emu/shadPS4), an in-early development PS4 emulator.
Closes #339617

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sh/shadps4/package.nix [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+  postPatch = ''
+    substituteInPlace src/common/path_util.cpp \
+      --replace-fail 'const auto user_dir = std::filesystem::current_path() / PORTABLE_DIR;' \
+      'const auto user_dir = std::filesystem::path(getenv("HOME")) / ".config" / "shadPS4";'
+  '';
+
+  configurePhase = ''
+    cmake -B build -DCMAKE_BUILD_TYPE=DEBUG -DENABLE_QT_GUI=ON
+  '';
```
Comment (OPNA2608) [inline pkgs/by-name/sh/shadps4/package.nix:49]:
```suggestion
  cmakeFlags = [
    (lib.cmakeBool "ENABLE_QT_GUI" true)
  ];

  # Still in development, help with debugging
  cmakeBuildType = "Debug";
  dontStrip = true;
```

File: pkgs/by-name/sh/shadps4/package.nix [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  buildPhase = ''
+    cmake --build build --config Release --parallel
+  '';
+
```
Comment (OPNA2608) [inline pkgs/by-name/sh/shadps4/package.nix:54]: 1. The `cmake` setup hook already handles calling this.
2. Passing `--parallel` as-is just completely disregards `NIX_BUILD_CORES`, we really don't want that.

```suggestion
```

File: pkgs/by-name/sh/shadps4/package.nix [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  meta = {
+    description = "Early in development PS4 emulator";
+    homepage = "https://github.com/shadps4-emu/shadPS4";
+    changelog = "https://github.com/shadps4-emu/shadPS4/releases/tag/v.${finalAttrs.version}";
+    license = lib.licenses.gpl2Only;
+    maintainers = with lib.maintainers; [ ryand56 ];
+    mainProgram = "shadps4";
+    platforms = lib.platforms.linux;
```
Comment (OPNA2608) [inline pkgs/by-name/sh/shadps4/package.nix:77]: On aarch64-linux:
```
In file included from /build/source/src/core/libraries/libc/printf.h:65,
                 from /build/source/src/core/libraries/libc/libc_stdio.h:7,
                 from /build/source/src/core/libraries/libc/libc_stdio.cpp:7:
/build/source/src/core/libraries/libc/va_ctx.h:4:10: fatal error: xmmintrin.h: No such file or directory
    4 | #include <xmmintrin.h>
      |          ^~~~~~~~~~~~~
compilation terminated.
```

So at minimum:

```suggestion
    platforms = lib.intersectLists lib.platforms.linux lib.platforms.x86;
```

Though prolly `lib.platforms.x86_64`, I don't think it'll work on 32-bit x86.

**Review by OPNA2608 — CHANGES_REQUESTED** [review]

**Comment by OPNA2608** [comment]
Have you been able to launch anything with this? For me the GUI launches and I can adjust settings & install homebrew pkgs, but it just freezes & eventually segfaults on everything eboot I try to launch.

**Comment by ryand56** [comment]
> Have you been able to launch anything with this? For me the GUI launches and I can adjust settings & install homebrew pkgs, but it just freezes & eventually segfaults on everything eboot I try to launch.

I haven't got around to it yet, but this might just be an issue with shadPS4 itself.

**Comment by OPNA2608** [comment]
I've got a later commit building & working on some OpenOrbis samples, though there are some leftover warnings at configure time - unpackaged dependencies, and outdated/borked provided dependencies. I think we should resolve those issues first before getting this in.

Bildschirmfoto_2024-09-13_20-18-00
[0001-shadps4-Get-it-working.patch.txt](https://github.com/user-attachments/files/16997962/0001-shadps4-Get-it-working.patch.txt)

**Comment by ryand56** [comment]
Seems like #331649 exists for toml11, but zydis v5.0.0 doesn't exist 🤔 and not sure about vulkan memory allocator

**Comment by OPNA2608** [comment]
`zydis`: https://github.com/shadps4-emu/shadPS4/pull/546

> the `zydis` commit in externals is from an unreleased dev branch (with an incompatible change) which will ultimately be v5, so v4 library need to be rejected

Our `zydis` package has no dependencies aside from `cmake`, so prolly fine to drop it from `buildInputs` and use the vendored one for now.

`vulkan-memory-allocator`: AFAICT the CMake config module doesn't expose its version, so it fails to determine if it satisfies the version constraint. Need to apply https://github.com/GPUOpen-LibrariesAndSDKs/VulkanMemoryAllocator/commit/257138b8f5686ae84491a3df9f90a77d5660c3bd to our package.

`sdl3`: https://github.com/NixOS/nixpkgs/pull/326699

**Comment by luochen1990** [comment]
Just syncing some info: 0.3.0 is released :)

**Comment by henkery** [comment]
Tested this by building it myself, it seems to crash and exit without explanation unexpectedly when attempting to load a game which runs on the official build (CUSA01113)

**Comment by ryand56** [comment]
> Tested this by building it myself, it seems to crash and exit without explanation unexpectedly when attempting to load a game which runs on the official build (CUSA01113)

Could you send the logs here? I'm still waiting on SDL3 to be added

**Comment by henkery** [comment]
> Could you send the logs here? I'm still waiting on SDL3 to be added

The logs don't say very much as it's a segfault
```[nix-shell:~/projects/shad]$ shadps4 
[Loader] <Info> emulator.cpp:Emulator:56: Starting shadps4 emulator v0.3.0 
[Loader] <Info> emulator.cpp:Emulator:57: Revision GITDIR-NOTFOUND
[Loader] <Info> emulator.cpp:Emulator:58: Branch GIT-NOTFOUND
[Loader] <Info> emulator.cpp:Emulator:59: Description GIT-NOTFOUND
[Config] <Info> emulator.cpp:Emulator:61: General isNeo: false
[Config] <Info> emulator.cpp:Emulator:62: GPU isNullGpu: false
[Config] <Info> emulator.cpp:Emulator:63: GPU shouldDumpShaders: false
[Config] <Info> emulator.cpp:Emulator:64: GPU shouldDumpPM4: false
[Config] <Info> emulator.cpp:Emulator:65: GPU vblankDivider: 1
[Config] <Info> emulator.cpp:Emulator:66: Vulkan gpuId: -1
[Config] <Info> emulator.cpp:Emulator:67: Vulkan vkValidation: false
[Config] <Info> emulator.cpp:Emulator:68: Vulkan vkValidationSync: false
[Config] <Info> emulator.cpp:Emulator:69: Vulkan vkValidationGpu: false
[Config] <Info> emulator.cpp:Emulator:70: Vulkan rdocEnable: false
[Config] <Info> emulator.cpp:Emulator:71: Vulkan rdocMarkersEnable: false
[Config] <Info> emulator.cpp:Emulator:72: Vulkan crashDiagnostics: false
[Kernel.Vmm] <Info> address_space.cpp:Impl:352: System managed virtual memory region: 0x400000 - 0x7ffffbfff
[Kernel.Vmm] <Info> address_space.cpp:Impl:355: System reserved virtual memory region: 0x7ffffc000 - 0xfffffffff
[Kernel.Vmm] <Info> address_space.cpp:Impl:358: User virtual memory region: 0x1000000000 - 0x10fffffffff
[Kernel.Vmm] <Info> memory.cpp:SetupMemoryRegions:50: Configured memory regions: flexible size = 0x1c000000, direct size = 0x120000000
[Kernel.Vmm] <Info> memory.cpp:MemoryManager:35: Usable memory address space: 1087_GB
[Loader] <Info> emulator.cpp:Run:126: Game id: CUSA01113 Title: Gravity Rush™ Remastered
[Loader] <Info> emulator.cpp:Run:129: Fw: 0x4500000 App Version: 01.00
Segmentation fault (core dumped)
```

**Comment by expenses** [comment]
https://github.com/shadps4-emu/shadPS4/issues/1042 has been merged so that first patch can be removed. For Bloodborne specifically you probably want to include https://github.com/shadps4-emu/shadPS4/pull/1146, but shadps4 development is happening so quickly that it might make more sense not to include any patches with this derivation and let people use `overrideAttrs` to point to their own sources.

**Comment by expenses** [comment]
20241001_12h44m48s_grim

Managed to get things working to some degree! I'm using https://github.com/shadps4-emu/shadPS4/pull/1146 and two patches:

https://github.com/shadps4-emu/shadPS4/issues/1042
```diff
--- a/src/core/libraries/kernel/thread_management.cpp
+++ b/src/core/libraries/kernel/thread_management.cpp
@@ -1065,7 +1065,16 @@ ScePthread PThreadPool::Create() {
         }
     }

+#ifdef _WIN64
     auto* ret = new PthreadInternal{};
+#else
+    // TODO: Linux specific hack
+    static u8* hint_address = reinterpret_cast<u8*>(0x7FFFFC000ULL);
+    auto* ret = reinterpret_cast<PthreadInternal*>(
+        mmap(hint_address, sizeof(PthreadInternal), PROT_READ | PROT_WRITE,
+             MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED, -1, 0));
+    hint_address += Common::AlignUp(sizeof(PthreadInternal), 4_KB);
+#endif
     ret->is_free = false;
     ret->is_detached = false;
     ret->is_almost_done = false;
```

and this to get controls with work without a numpad:
```diff
diff --git a/src/sdl_window.cpp b/src/sdl_window.cpp
index bd2cc39d..13438149 100644
--- a/src/sdl_window.cpp
+++ b/src/sdl_window.cpp
@@ -127,19 +127,12 @@ void WindowSDL::onResize() {
 void WindowSDL::onKeyPress(const SDL_Event* event) {
     using Libraries::Pad::OrbisPadButtonDataOffset;

-#ifdef __APPLE__
     // Use keys that are more friendly for keyboards without a keypad.
     // Once there are key binding options this won't be necessary.
     constexpr SDL_Keycode CrossKey = SDLK_N;
     constexpr SDL_Keycode CircleKey = SDLK_B;
     constexpr SDL_Keycode SquareKey = SDLK_V;
     constexpr SDL_Keycode TriangleKey = SDLK_C;
-#else
-    constexpr SDL_Keycode CrossKey = SDLK_KP_2;
-    constexpr SDL_Keycode CircleKey = SDLK_KP_6;
-    constexpr SDL_Keycode SquareKey = SDLK_KP_4;
-    constexpr SDL_Keycode TriangleKey = SDLK_KP_8;
-#endif

     u32 button = 0;
     Input::Axis axis = Input::Axis::AxisMax;
```

and then using this for compilation:
```nix
with (import /home/ashley/projects/nixpkgs { });
shadps4.overrideAttrs (_: {
  src = lib.cleanSource ./.;
  patches = [ ./bloodborne.patch ./laptop-controls.patch ];
  cmakeBuildType = "RelWithDebugInfo";
})
```

It doesn't work super well with integrated graphics but still!

>Tested this by building it myself, it seems to crash and exit without explanation unexpectedly when attempting to load a game which runs on the official build (CUSA01113)

I also had this and ran `gdb` which shows a trace related to it being unable to load the correct X11 libs. I reused a devshell from a previous project in order to get that stuff to work:

```nix
devShells.default =
          with pkgs;
          mkShell rec {
            nativeBuildInputs = [
              pkg-config
              trunk
              wasm-bindgen-cli
            ];
            buildInputs = [
              udev
              alsa-lib
              vulkan-loader
              xorg.libX11
              xorg.libXcursor
              xorg.libXi
              xorg.libXrandr # To use the x11 feature
              libxkbcommon
              wayland # To use the wayland feature
            ];
            LD_LIBRARY_PATH = lib.makeLibraryPath buildInputs;
          };
```

**Comment by expenses** [comment]
> @henkery I also had this and ran `gdb` which shows a trace related to it being unable to load the correct X11 libs. I reused a devshell from a previous project in order to get that stuff to work:

```nix
shad2 = with pkgs; writeShellApplication {
            name = "shad2";
            text = ''${shadps4}/bin/shadps4 "$@"'';
            runtimeEnv.LD_LIBRARY_PATH = lib.makeLibraryPath [
              xorg.libXi

            ];
          };
```

this works equally well.

**Comment by OPNA2608** [comment]
Prolly needs smth like this:

https://github.com/NixOS/nixpkgs/blob/d78d09350ac7dfe503cf48cbc59764aef4157b9a/pkgs/development/libraries/SDL2/default.nix#L206-L218

...but for `$out/bin` & just whatever X11/wayland libs it wants to `dlopen()`.

**Comment by expenses** [comment]
> Prolly needs smth like this:

Yep, I was just testing using a wrapper script as it didn't require recompilation, but you can just patch the rpaty in the binary. See my edited comment.

**Comment by expenses** [comment]
I've uploaded my changes at https://github.com/NixOS/nixpkgs/compare/master...expenses:nixpkgs:shadps4. I'm using a local path for the shadps4 source but that's only because I'm on a bad internet connection and it cuts off before I can clone the repo in nix. I'm using the latest head (https://github.com/shadps4-emu/shadPS4/commit/bf3e43b01683bd783323e12237027103fae28d54).

**Comment by expenses** [comment]
Things look good now :)

**Comment by urioTV** [comment]
I just want to say that it seems working.

**Comment by pinknoseboopy** [comment]
The emulator crashed when clicking play, but it may not be due to the PR since the appimage also crashes. The errors show that the emulator doesn't support certain thread settings, has some missing functions, and can't fully access the game files. https://github.com/shadps4-emu/shadPS4/issues/1311

**Review by henkery — APPROVED** [review]
It's working well and as expected
perhaps we wish to discuss the release/update cadence here as well
I see we've now switched to unstable versions but how will we maintain which version we track, is it just the maintainer who updates the tracked commit whenever they feel like it?

**Comment by ryand56** [comment]
> It's working well and as expected perhaps we wish to discuss the release/update cadence here as well I see we've now switched to unstable versions but how will we maintain which version we track, is it just the maintainer who updates the tracked commit whenever they feel like it?

I'm depending on the unstable releases for now, until stable 0.3.1 is released.

**Comment by urioTV** [comment]
We can always depend on latest release. For bleeding edge there is always https://github.com/chaotic-cx/nyx repo where bot automatically updates to the latest git commit.

**Comment by OPNA2608** [comment]
If you want, you can add an update script.

```nix
passthru.updateScript = unstableGitUpdater {
  tagPrefix = "v";
};
```

Should make it trivial for anyone to make a bump (`nix-shell maintainers/scripts/update.nix --argstr package shadps4`), and @r-ryantm will automatically generate bump PRs whenever it runs the update script and the bumped build passes (which can happen *abit too frequently* on unstable packages with active upstreams like this, so should maybe [opt out of the auto-updates in case the spam bothers you](https://github.com/NixOS/nixpkgs/blob/80792e6e0103e99fd2f1d53e32dc7c90628352d1/pkgs/development/compilers/open-watcom/v2.nix#L15)).

(I'll give this another review later today, not on an x86_64 machine right now)

**Comment by OPNA2608** [comment]
New auto-updater doesn't make sense for our purposes & might confuse users with it non-functionality on NixOS: https://github.com/shadps4-emu/shadPS4/issues/1368

Patch to neutralise it for now:

[0001-Disable-update-checking.patch.txt](https://github.com/user-attachments/files/17357353/0001-Disable-update-checking.patch.txt)

image

(didn't wanna dig deeper into how to make Qt not get stuck on weirdly initiated windows on app shutdown, otherwise I'd make it throw a warning & immediately exit [here](https://github.com/shadps4-emu/shadPS4/blob/bd9f82df94847b4a5f3d2676ae938f064505c992/src/qt_gui/check_update.cpp#L41) instead of hijacking that dialogue)

File: pkgs/by-name/sh/shadps4/package.nix [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        vulkan-loader
+        xorg.libXi
+      ]
+    } \
+      $out/bin/shadps4
+  '';
+
+  passthru.updateScript = unstableGitUpdater {
+    tagPrefix = "v";
+  };
```
Comment (OPNA2608) [inline pkgs/by-name/sh/shadps4/package.nix:129]: Gah.

1. Misread the tag format, they're using `v.X.Y.Z` (note the extra `.` immediately after the v)
2. *This* popped up earlier today:
  image

So:

```suggestion
  passthru.updateScript = unstableGitUpdater {
    tagFormat = "v.*";
    tagPrefix = "v.";
  };
```

File: pkgs/by-name/sh/shadps4/package.nix [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  vulkan-memory-allocator,
+  xorg,
+  xxHash,
+  zlib-ng,
+  unstableGitUpdater,
+}:
+
+stdenv.mkDerivation {
+  pname = "shadps4";
+  version = "0.3.1-unstable-2024-10-13";
```
Comment (OPNA2608) [inline pkgs/by-name/sh/shadps4/package.nix:38]: With fixed update script, it disagrees on the version:

```
+ update-source-version shadps4 0.3.0-unstable-2024-10-13 --rev=bd9f82df94847b4a5f3d2676ae938f064505c992
update-source-version: error: Failed to replace source revision 'bd9f82df94847b4a5f3d2676ae938f064505c992' to 'bd9f82df94847b4a5f3d2676ae938f064505c992' in 'shadps4'!
```

Base version should be the latest released version, so [0.3.0](https://github.com/shadps4-emu/shadPS4/releases/tag/v.0.3.0).

```suggestion
  version = "0.3.0-unstable-2024-10-13";
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]

File: pkgs/by-name/sh/shadps4/package.nix [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+  passthru.updateScript = unstableGitUpdater {
+    tagFormat = "v.*";
+    tagPrefix = "v.";
+  };
+
+  meta = {
+    description = "Early in development PS4 emulator";
+    homepage = "https://github.com/shadps4-emu/shadPS4";
+    license = lib.licenses.gpl2Only;
```
Comment (OPNA2608) [inline pkgs/by-name/sh/shadps4/package.nix:135]: Source seems to have thorough (although not perfectly consistent) license annotations, so:

```
↪ grep -ri 'SPDX-License-Identifier' | cut -d':' -f2- | sort | uniq
SPDX-License-Identifier = "BSD-3-Clause"
// SPDX-License-Identifier: BSD-3-Clause AND GPL-2.0-or-later
SPDX-License-Identifier = "BSL-1.0"
// SPDX-License-Identifier: CC0-1.0
SPDX-License-Identifier = "CC0-1.0"
             SPDX-License-Identifier: GPL-2.0-or-later -->
         SPDX-License-Identifier: GPL-2.0-or-later -->
     SPDX-License-Identifier: GPL-2.0-or-later -->
# SPDX-License-Identifier: GPL-2.0-or-later
//  SPDX-License-Identifier: GPL-2.0-or-later
// SPDX-License-Identifier: GPL-2.0-or-later
// SPDX-License-Identifier: GPL-2.0-or-later
SPDX-License-Identifier = "GPL-2.0-or-later"
SPDX-License-Identifier: GPL-2.0-or-later
//  SPDX-License-Identifier: MIT
SPDX-License-Identifier = "MIT"
SPDX-License-Identifier = "OFL-1.1"
```

All `GPL-2.0` occurrences are also `-or-later`, so `gpl2Plus` is *more* correct I think.

```suggestion
    license = lib.licenses.gpl2Plus;
```

**Review by OPNA2608 — CHANGES_REQUESTED** [review]
Last thing I've spotted, otherwise LGTM.

**Review by OPNA2608 — APPROVED** [review]
Result of `nixpkgs-review pr 340215` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>shadps4</li>
  </ul>
</details>

---

Builds on OfBorg, OpenOrbis sample package still works, can't spot anything immediately wrong with it anymore. LGTM!

**Pull request merged**